PYIC-4258: Validate and store inherited identity with new CriConfig

MikeCollingwood opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

- Added new class, `CriConfig`, to be able to validate the signed inherited identity JWT
- Made configs extend this new class instead of implement interface.
- Renamed these configs to OauthCriConfig and RestConfig.
- Added in functionality to validate and store inheritedIdentityJWT, behind a feature flag

### Why did it change

- We need to be able to process inherited identities received from the RP as part of the request claims set

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-4258](https://govukverify.atlassian.net/browse/PYIC-4258)

See this PR for changes to parameters and the addition of the feature flag:
https://github.com/govuk-one-login/ipv-core-common-infra/pull/819

In theory, this sh/could work straight out of the box, though it needs manual testing. This could be done by deploying `orch-stub` hardcoded to respond with the inherited identity. Seeing the VC in DynamoDB would prove success.

Talat is working on the stub, which would be a great way to test this code. In addition, George Nemeshanyi is working on giving us the inherited identity JWT from orchestration, probably good to talk to them at some point!

Another thing, Sonar is breaking checks. Something to check with an SRE because looks like it could be an env problem.

[PYIC-4258]: https://govukverify.atlassian.net/browse/PYIC-4258?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

MikeCollingwood committed:
PYIC-4258: validate and store inherited identity, with new CriConfig

MikeCollingwood commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/dto/CriConfig.java:
```diff
@@ -1,14 +1,23 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@SuperBuilder
+public class CriConfig {
```
Classes which previously implemented the CriConfig interface now just extend this class. This is useful because this class is what is needed for validating the inheritedIdentityJWT

The classes which now extend this class were renamed for clarity:
- "CredentialIssuerConfig" => "OauthCriConfig"
- "BackendCriConfig" => "RestCriConfig"

MikeCollingwood committed:
PYIC-4258: Add unit test to check it validates and stores

MikeCollingwood committed:
PYIC-4258: Refactor code and remove comments

MikeCollingwood commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -155,6 +171,15 @@ public APIGatewayProxyResponseEvent handleRequest(
                             govukSigninJourneyId,
                             ipAddress);
 
+            if (configService.enabled(CoreFeatureFlag.INHERITED_IDENTITY)) {
```
Here's the important stuff!

Behind a feature flag it tries to validate and story any inherited identities.

The userId passed in comes from the claimSet sub of the parent VC. When validating the inherited identity jwt, we check the subject is equal to this id. This should reassure you that we're storing it under the correct userId!

MikeCollingwood committed:
PYIC-4258: Fix test getInheritedIdentityJWT userId choice

Wynndow commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -224,9 +249,26 @@ public APIGatewayProxyResponseEvent handleRequest(
+                signedInheritedIdentityJWT, inheritedIdentityCriConfig, userId);
+
+        verifiableCredentialService.persistUserCredentials(
+                signedInheritedIdentityJWT, HMRC_MIGRATION_CRI, userId);
```
I'm going to hazard a guess we're going to need a new audit event in here.

Joe-Edwards-GDS replied:
Audit event is covered by https://govukverify.atlassian.net/browse/PYIC-4248 (or perhaps one of the associated tickets)

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -224,9 +249,26 @@ public APIGatewayProxyResponseEvent handleRequest(
         }
     }
 
+    private void validateAndStoreHMRCInheritedIdentity(String inheritedIdentityJWT, String userId)
```
We'll also need to include in the `vcReceivedThisSession` array (for TICF)

Wynndow replied:
Yep, good spot.

This raises an interesting question though. A user comes in on a migration journey with an inherited identity VC with a `vot` of `PCL250` and a `vtr` including `PCL250`. In theory we happily match the operational profile and send the user back to the service as a successful journey.

Before we issue the identity we'll send the inherited identity VC to TICF. What happens if TICF responds with a VC containing a CI? Do we fail the journey? Because presumably the identity is already issued and has no concept of CIs. It may be that TICF would never issue a CI in this scenario? Or do we ignore the response from TICF in that case?

Joe-Edwards-GDS replied:
In general we respect CIs as normal, so in this case we'd fail the journey - TICF may be using other information to make that judgement (not sure what exactly...), but you're probably right that it's unlikely :)

Wynndow replied:
I guess I'm wondering if that's acceptable to HMRC? A user may be blocked using our system that previously wasn't using their system.

Joe-Edwards-GDS replied:
Yep - the latest direction is that if we've identified breaching CIs, then HMRC would like to know about it.

(They went back and forth on this a bit - as you say, it will block access to some existing users)

Wynndow committed:
PYIC-4258: Handle inheritedIdentityJwt Failures

If there is an issue with the received inherited identity jwt - badly
formatted, or doesn't validate - we don't need to just return a technical
error. By the point we're throwing an error we'll have already validated
the clients redirect URI. That means we can close the OAuth loop and
send the user back.

This also adds domain objects to parse the claims/userinfo json blob
into. It looks like we won't be receiving the inhertied identity jwt on
the top level, but rather with the other requested claims (which we
currently ignore).

Wynndow committed:
PYIC-4258: Fix equals methods across CRI hierarchy

This gives each level of the CRI config a fully working equals method.
They can be used in our tests to simplify things a bit.

It moves the getter for the parsed keys to a method that doesn't clash
with the default getter - this allows the equals methods to work.

Wynndow committed:
PYIC-4258: Allow no inherited identity claim

Most of the time we won't be receiving an inherited identity jwt, so we
need to allow this to happen once the feature flag is removed.

This checks the claims claim and the userinfo object and the inherited
identity claim. If any of them are null we return.

In reality we should always be getting a claims claim and a userinfo
claim, as they're currently in the JAR from orch, but should still do
the null checks when diving into this object. Perhaps we should be
throwing though if they are null.

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -249,24 +246,53 @@ public APIGatewayProxyResponseEvent handleRequest(
+        try {
+            signedInheritedIdentityJWT =
+                    SignedJWT.parse(
+                            ((JarClaims) claimsSet.getClaim("claims"))
```
I'm not convinced this will work - I think you need to explicitly map this as a JSON value (you can't just cast it).

Wynndow replied:
Yep, you're quite right. The tests are all constructing a claimsSet with the right objects in which is why the tests pass. Need to make sure it's been deserialized/serialized to be a valid test.

Wynndow committed:
PYIC-4258: Store inherited identity VC for TICF

We need to send the inherited identity VC to TICF, so this adds it to
the maintained list of VC received in the users session.

Wynndow committed:
PYIC-4258: Correctly convert claims claim

This change ensure that we're correctly converting the values in the
claims claim. Previously we were just casting them. This worked as the
test values were being set directly in the claimsset object and not
undergoing any deserialization/serialization.

This updates the tests to ensure that the value are the same as those
we'll receive over the wire. They are actually presented to us as
JSONObjects, which we can convert into the domain objects we can work
with.

Wynndow committed:
PYIC-4258: Tidy up JarValidatorTest tests

There was a lot of repition which it's nice to dry up.

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -227,6 +251,76 @@ public APIGatewayProxyResponseEvent handleRequest(
+        var inheritedIdentityJwtClaim = userInfo.inheritedIdentityClaim();
+        if (inheritedIdentityJwtClaim == null) {
+            return;
+        }
```
Can we separate this logic into a different method (which will make it much more reusable if we introduce other inherited ids) - i.e. separate `getInheritedIdentityClaim` and `validateAndStoreHMRCInheritedIdentity`?

Joe-Edwards-GDS replied:
These null checks are pretty ugly :( - shame Java doesn't let you do conditional access.

One slightly sneaky way of simplifying is leveraging optionals, for example:
```
var inheritedIdClaim = Optional.ofNullable(claims)
        .map(JarClaims::userInfo)
        .map(JarUserInfo::inheritedIdentityClaim)
        .map(InheritedIdentityJwtClaim::value)
        .flatMap((values) -> values.stream().findFirst());
```

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -227,6 +251,76 @@ public APIGatewayProxyResponseEvent handleRequest(
+        }
+    }
+
+    private RecoverableJarValidationException buildRecoverableException(
```
So this exception will (after some back and forth) result in the user going straight back to the RP with an error?

Joe-Edwards-GDS replied:
Do you think this should be a constructor for `RecoverableJarValidationException` (or factory method if the exception is a problem)?

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -227,6 +251,76 @@ public APIGatewayProxyResponseEvent handleRequest(
+            return;
+        }
+
+        SignedJWT signedInheritedIdentityJWT;
```
Whenever I see this pattern it smells of needing to extract a method - if you have a method, you can return from within the block.

(In truth, I just hate any non-final variables :) )

MikeCollingwood replied:
I've made it final by nesting a try/ catch, though could do this another way

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -227,6 +251,76 @@ public APIGatewayProxyResponseEvent handleRequest(
+            verifiableCredentialService.persistUserCredentials(
+                    signedInheritedIdentityJWT, HMRC_MIGRATION_CRI, userId);
+            ipvSessionItem.setVcReceivedThisSession(
+                    List.of(signedInheritedIdentityJWT.serialize()));
```
I wonder if we should add an `addVcReceivedThisSession` method to `IpvSessionItem` - I think it might simplify the code in the other use-case as well...

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/validation/JarValidator.java:
```diff
@@ -169,11 +173,13 @@ private JWTClaimsSet getValidatedClaimSet(SignedJWT signedJWT, String clientId)
 
-            validateMaxAllowedJarTtl(signedJWT.getJWTClaimsSet());
+            validateMaxAllowedJarTtl(claimsSet);
+            validateInheritedIdentityJwtClaim(claimsSet);
```
Can we put this behind the feature flag, to avoid the risk of breaking any prod journeys prematurely?

MikeCollingwood replied:
Have moved validation to only be in `validateAndStoreHMRCInheritedIdentity`

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/validation/JarValidator.java:
```diff
@@ -223,4 +229,44 @@ private URI validateRedirectUri(JWTClaimsSet claimsSet, String clientId)
         }
     }
+
+    private void validateInheritedIdentityJwtClaim(JWTClaimsSet claimsSet)
```
It feels a bit unsatisfactory to convert the claims object and perform almost all the same checks twice, but it looks ugly to avoid :(

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/test/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandlerTest.java:
```diff
@@ -80,33 +116,50 @@ class InitialiseIpvSessionHandlerTest {
+        jarClaims.put(VTR, List.of("Cl.Cm.P2", "Cl.Cm.PCL200"));
+        jarClaims.put(
+                CLAIMS,
+                new JarClaims(
```
Is this allowed - can we put a custom class here? I assume it works, but strictly speaking I think these should be maps.

MikeCollingwood replied:
Is there a reason we wouldn't want to build it like this:
```
         JWTClaimsSet claimsSet =
                new JWTClaimsSet.Builder()
                        .expirationTime(new Date(Instant.now().plusSeconds(1000).getEpochSecond()))
                        .issueTime(new Date())
                        .notBeforeTime(new Date())
                        .subject("test-user-id")
                        .audience("test-audience")
                        .issuer("test-issuer")
                        .claim("response_type", "code")
                        .claim("redirect_uri", "http://example.com")
                        .claim("state", "test-state")
                        .claim("client_id", "test-client")
                        .claim("vtr", List.of("Cl.Cm.P2", "Cl.Cm.PCL200"))
                        .build();
```

Joe-Edwards-GDS replied:
I think it's because there are some tests that might modify these. It might be neater to have a 'default' builder which can be appended to somehow. I haven't thought through all the details though...

MikeCollingwood committed:
PYIC-4258: Refactor InitialiseIpvSessionHandler:
- Remove inheritedIdentityJwt Jar validation at top level
- Refactor inheritedIdentityJwt validation
- Add addVcReceivedThisSession method to IpvSessionItem (replacing both setVcReceivedThisSession usages)
- Separate a getInheritedIdentityClaim method to simplify check for claim before calling validate and store method

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -227,6 +258,88 @@ public APIGatewayProxyResponseEvent handleRequest(
+                    claimsSet);
+        } catch (JarValidationException e) {
+            throw buildRecoverableException(e.getErrorObject(), claimsSet);
+        }
```
If you separate this into two methods, I think you can avoid the nested try/catch? i.e. `validateHmrcInheritedIdentity` and `storeHmrcInheritedIdentity` (the 'and' in this method name hints at it too!)

Joe-Edwards-GDS replied:
(With some subtlety to distinguish validating that the JAR claims structure is valid, and that the VC contained in the claim is valid)

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -227,6 +258,88 @@ public APIGatewayProxyResponseEvent handleRequest(
+                                    inheritedIdentityJwtList.size())));
+        }
+        return inheritedIdentityJwtList;
+    }
```
Given that this method validates that the list contains exactly one element, should it just return that directly?

rpayal commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -227,6 +258,88 @@ public APIGatewayProxyResponseEvent handleRequest(
+
+            // Store
+            try {
+                verifiableCredentialService.persistUserCredentials(
```
When we r persisting the migrated hmrc VC........let say the req was for vtr PCL250. 
Do we here also need to check for any previous hmrc vc for vtr PCL200 (i.e all lower profile vc's) and delete it ?

MikeCollingwood replied:
Think we have a [ticket](https://govukverify.atlassian.net/browse/PYIC-4384https://govukverify.atlassian.net/browse/PYIC-4384) for looking into this

MikeCollingwood committed:
PYIC-4258: Refactor RecoverableJarValidationException:
- Add constructor for RecoverableJarValidationException using claimsSet
- Simplify RecoverableJarValidationException and JarValidationException with Lombok annotations

MikeCollingwood committed:
PYIC-4258: Split validating and storing inherited identity methods

MikeCollingwood committed:
PYIC-4258: Simplify test setup

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -231,6 +261,97 @@ public APIGatewayProxyResponseEvent handleRequest(
+            throws JarValidationException, ParseException, VerifiableCredentialException {
+        // Validate JAR claims structure is valid
+        var inheritedIdentityJwtList =
+                Optional.of(inheritedIdentityJwtClaim.value())
```
You need to use `ofNullable` if you think it may be nullable (`of` will throw an NPE)

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -288,30 +295,30 @@ private void validateAndStoreHMRCInheritedIdentity(
             verifiableCredentialJwtValidator.validate(
                     signedInheritedIdentityJWT, inheritedIdentityCriConfig, userId);
-            LogHelper.logMessage(Level.INFO, "Migration VC successfully validated");
+            LOGGER.info("Migration VC successfully validated");
```
Can we use `LogHelper.buildLogMessage` here?

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/domain/InheritedIdentityJwtClaim.java:
```diff
@@ -0,0 +1,5 @@
+
+import java.util.List;
+
+public record InheritedIdentityJwtClaim(List<String> value) {}
```
Sorry Mike - this has just been changed to `values` in the RFC 🤦 

(kinda makes sense, given it's a list...)

https://github.com/govuk-one-login/architecture/pull/554/files

Joe-Edwards-GDS reviewed (commented):
I think this is almost there!

MikeCollingwood committed:
PYIC-4258: Format with spotlessApply

MikeCollingwood committed:
PYIC-4258: Make inherited identities value attribute plural

Joe-Edwards-GDS reviewed (approved):
I think this is good, thanks, and we can come back to it as part of the other tickets if anything comes up. :)
